File: ex17_stack.py
# ex17_stack.py
# run this python3
# This short program implements and uses the Stack data structure.

# The Stack class
class Stack:
    def __init__(self):
        self.__storage = []

    def isEmpty(self):
        return len(self.__storage) == 0
    
    def push(self,p):
        self.__storage.append(p)
    
    def pop(self):
        return self.__storage.pop()


# Now, let's use the Stack class!!!

theStack = Stack()

# First, push some stuff onto the stack
print("\nEnter values to push onto the stack.\n")
for i in range(0, 10):
    value = input("Enter a value: ")
    theStack.push(value)

# Second, pop the stuff off the stack
print("\nPopping the values from the stack.\n")
while not theStack.isEmpty():
    print("Value: %r" % theStack.pop())

File: ex17_stack.c
/* ex17_stack.c
   This is a short program that implements and demonstrates
   a linked list implementaiton of a stack data structure.
*/

#include <stdio.h>
#include <stdlib.h>

struct Node {
  struct Node *next;
  int data;
};

struct Stack {
  struct Node *top;
};


struct Stack* Stack_new () {
  struct Stack *stack = malloc(sizeof(struct Stack));
  stack->top = NULL;
  return stack;
}


int Stack_isEmpty (struct Stack *stack) {
  return (stack->top == NULL);
}


void Stack_push (struct Stack *stack, int value) {
  struct Node *node = malloc(sizeof(struct Node));

  node->data = value;
  node->next = stack->top;
  stack->top = node;
}


int Stack_pop (struct Stack *stack) {
  int value = 0;
  struct Node *temp;

  value = stack->top->data;
  temp = stack->top;
  stack->top = stack->top->next;
  free(temp);

  return value;
}


void Stack_delete (struct Stack *stack) {
  while (!Stack_isEmpty(stack)) {
    Stack_pop(stack);
  }
  
  free(stack);
}



int main (int argc, char** argv) {
  struct Stack *theStack;
  
  theStack = Stack_new();
  
  printf("\nPushing 0 through 9 on to the stack.\n");
  int i = 0;
  for (i = 0; i < 10; i++) {
    Stack_push(theStack, i);
  }
  
  printf("\nPopping the values from the stack.\n");
  int value = 0;
  while (!Stack_isEmpty(theStack)) {
    value = Stack_pop(theStack);
    printf("Value: %d\n", value);
  }
  
  Stack_delete(theStack);
  
  return 0;
}
